erpnext/accounts/doctype/payment_request/payment_request.py

```diff
@@ -164,53 +163,48 @@ def before_submit(self):
 		elif self.payment_request_type == "Inward":
 			self.status = "Requested"
 
-		if self.payment_request_type == "Inward":
-			if self.payment_channel == "Phone":
-				self.request_phone_payment()
-			else:
-				self.set_payment_request_url()
+		if self.payment_request_type == "Inward" and self.payment_gateway:
+			controller = _get_payment_gateway_controller(self.payment_gateway)
+			tx_data = self.get_tx_data()
+			controller.on_refdoc_submission(tx_data)  # preflight check
+			self.integration_request = controller.initiate_payment(tx_data)
+			# checkout page can also be the target of a gatway initialized flow
+			# gateways also may return None if appropriate
+			self.payment_url = controller.get_payment_url(self.integration_request)
+			if not controller.is_user_flow_initiation_delegated(self.integration_request):
 				if not (self.mute_email or self.flags.mute_email):
 					self.send_email()
 					self.make_communication_entry()
 
-	def request_phone_payment(self):
-		controller = _get_payment_gateway_controller(self.payment_gateway)
-		request_amount = self.get_request_amount()
-
-		payment_record = dict(
-			reference_doctype="Payment Request",
-			reference_docname=self.name,
-			payment_reference=self.reference_name,
-			request_amount=request_amount,
-			sender=self.email_to,
-			currency=self.currency,
-			payment_gateway=self.payment_gateway,
-		)
-
-		controller.validate_transaction_currency(self.currency)
-		controller.request_for_payment(**payment_record)
+	def on_cancel(self):
+		self.check_if_payment_entry_exists()
+		self.set_as_cancelled()
 
-	def get_request_amount(self):
-		data_of_completed_requests = frappe.get_all(
-			"Integration Request",
-			filters={
+	def get_tx_data(self):
+		party = frappe.get_doc(self.party_type, self.party)
+		if self.party_type == "Supplier" and party.supplier_primary_contact:
+			contact = frappe.get_doc("Contact", party.supplier_primary_contact).as_dict()
+		elif self.party_type == "Customer" and party.customer_primary_contact:
+			contact = frappe.get_doc("Contact", party.customer_primary_contact).as_dict()
+		else:
+			contact = {}
+		if self.party_type == "Supplier" and party.supplier_primary_address:
+			address = frappe.get_doc("address", party.supplier_primary_address).as_dict()
+		elif self.party_type == "Customer" and party.customer_primary_address:
+			address = frappe.get_doc("address", party.customer_primary_address).as_dict()
+		else:
+			address = {}
+		return frappe._dict(
+			{
+				"amount": self.grand_total,
+				"currency": self.currency,
 				"reference_doctype": self.doctype,
 				"reference_docname": self.name,
-				"status": "Completed",
-			},
-			pluck="data",
+				"payer_contact": contact,
+				"payer_address": address,
+			}
```

Comment:
Should probably go into the transaction controller for a generic proto-payment-gateway support... :thinking: